src/pudl/extract/ferc1.py

```diff
@@ -755,33 +768,26 @@ def extract_dbf(
 
     ferc1_raw_dfs = {}
     for pudl_table in ferc1_settings.tables:
+        logger.info(
+            f"Converting extracted FERC Form 1 table {pudl_table} into a "
+            f"pandas DataFrame from DBF table."
+        )
         if pudl_table not in TABLE_NAME_MAP_FERC1:
             raise ValueError(
                 f"No extract function found for requested FERC Form 1 data "
                 f"table {pudl_table}!"
             )
-        logger.info(
-            f"Converting extracted FERC Form 1 table {pudl_table} into a "
-            f"pandas DataFrame from DBF table."
-        )
-        if pudl_table == "income_statement_ferc1":
-            # special case for the income statement. bc the dbf table is two tables.
-            income_tbls = []
-            for raw_income_table_name in TABLE_NAME_MAP_FERC1[pudl_table]["dbf"]:
-                income_tbls.append(
-                    extract_dbf_generic(
-                        ferc1_engine=sa.create_engine(pudl_settings["ferc1_db"]),
-                        ferc1_settings=ferc1_settings,
-                        table_name=raw_income_table_name,
-                    ).assign(sched_table_name=raw_income_table_name)
-                )
-            ferc1_raw_dfs[pudl_table] = pd.concat(income_tbls)
+        dbf_table_or_tables = TABLE_NAME_MAP_FERC1[pudl_table]["dbf"]
+        if not isinstance(dbf_table_or_tables, list):
+            dbf_tables = [dbf_table_or_tables]
         else:
-            ferc1_raw_dfs[pudl_table] = extract_dbf_generic(
-                ferc1_engine=sa.create_engine(pudl_settings["ferc1_db"]),
-                ferc1_settings=ferc1_settings,
-                table_name=TABLE_NAME_MAP_FERC1[pudl_table]["dbf"],
-            )
+            dbf_tables = dbf_table_or_tables
+
+        ferc1_raw_dfs[pudl_table] = extract_dbf_generic(
```

Comment:
non-blocking: I guess you could build this ferc1_raw_dfs as part of a comprehension too, if you wanted - though that may be more annoying than it's worth.

On this note, I realized a lot of my structure comments have been about "how to clean up gnarly for loops." [This presentation](https://www.dabeaz.com/generators/Generators.pdf) was really influential in my own thinking about this - especially slides 30-46 I think are relevant. IDK if it will hit you the same but I was like "woah, if you get the function input/outputs right you can get a very nice-looking pipeline for just about anything!"

Reply:
Thank you! I will definitely dive into that presentation.

Reply:
> non-blocking: I guess you could build this ferc1_raw_dfs as part of a comprehension too, if you wanted - though that may be more annoying than it's worth.

Can you show me what you mean?

Reply:
Yeah, sorry I didn't do that earlier! Basically taking the code you have and breaking it into smaller functions. I put them inline because I think that makes the diff clearer but you'd probably at least want the 'munge_to_list' to live outside of the function body itself.

```diff
diff --git a/src/pudl/extract/ferc1.py b/src/pudl/extract/ferc1.py
index f7e7c1a1..637422e4 100644
--- a/src/pudl/extract/ferc1.py
+++ b/src/pudl/extract/ferc1.py
@@ -817,40 +817,36 @@ def extract_xbrl(
     if ferc1_settings is None:
         ferc1_settings = Ferc1Settings()
 
-    ferc1_raw_dfs = {}
     if not ferc1_settings.xbrl_years:
-        return ferc1_raw_dfs
-
-    for pudl_table in ferc1_settings.tables:
-        if pudl_table not in TABLE_NAME_MAP_FERC1:
-            raise ValueError(f"{pudl_table} not found in the list of known tables.")
-        if "xbrl" not in TABLE_NAME_MAP_FERC1[pudl_table]:
-            raise ValueError(f"No XBRL tables have been associated with {pudl_table}.")
-
-        logger.info(
-            f"Converting extracted FERC Form 1 table {pudl_table} into a "
-            f"pandas DataFrame from XBRL table."
-        )
+        return {}
 
-        xbrl_table_or_tables = TABLE_NAME_MAP_FERC1[pudl_table]["xbrl"]
-        if not isinstance(xbrl_table_or_tables, list):
-            xbrl_tables = [xbrl_table_or_tables]
-        else:
-            xbrl_tables = xbrl_table_or_tables
-
-        ferc1_xbrl_engine = sa.create_engine(pudl_settings["ferc1_xbrl_db"])
-
-        ferc1_raw_dfs[pudl_table] = {
+    def get_periods(xbrls, engine):
+        return {
             period_type: extract_xbrl_generic(
-                table_names=xbrl_tables,
+                table_names=xbrls,
                 period=period_type,
-                ferc1_engine=ferc1_xbrl_engine,
+                ferc1_engine=engine,
                 ferc1_settings=ferc1_settings,
             )
             for period_type in ["duration", "instant"]
         }
 
-    return ferc1_raw_dfs
+    def munge_to_list(item_or_items):
+        if isinstance(item_or_items, list):
+            return item_or_items
+        return [item_or_items]
+
+    ferc1_xbrl_engine = sa.create_engine(pudl_settings["ferc1_xbrl_db"])
+
+    xbrls_to_extract = (
+        (pudl_name, munge_to_list(xbrl))
+        for pudl_name in ferc1_settings.tables
+        if (xbrl := TABLE_NAME_MAP_FERC1.get(pudl_name, {}).get("xbrl")) is not None
+    )
+    return {
+        pudl_table: get_periods(xbrls, ferc1_xbrl_engine)
+        for pudl_table, xbrls in xbrls_to_extract
+    }
 
 
 def extract_xbrl_single(
```

Reply:
This logic is cool to review / I will hopefully use stuff like this in the future, but I wonder what the general consensus is on for clarity vs. efficiency. I feel like it's good practice to pull out small tasks as individual functions, but also sometimes more abstracting of tasks makes them harder to follow? Maybe it's just my brain hurting after all this sales table stuff...